src/server/stream_family.cc

```diff
@@ -1402,44 +1533,131 @@ std::optional<ReadOpts> ParseReadArgsOrReply(CmdArgList args, ConnectionContext*
     return std::nullopt;
   }
 
+  vector<unordered_map<string_view, StreamIDsItem>> res_sitems(shard_set->size());
+  bool stop = false;
+  auto g_cb = [&](Transaction* t, EngineShard* shard) {
+    if (!opts.read_group) {
+      return OpStatus::OK;
+    }
+    auto sid = shard->shard_id();
+    auto s_args = t->GetShardArgs(sid);
+    unordered_map<string_view, StreamIDsItem> sid_items;
+
+    // get group and consumer
+    for (size_t i = 0; i < s_args.size(); i++) {
+      string_view key = s_args[i];
+
+      streamCG* group = nullptr;
+      streamConsumer* consumer = nullptr;
+      auto group_res = FindGroup(t->GetOpArgs(shard), key, opts.gname);
+      if (!group_res) {
+        (*cntx)->SendError(group_res.status());
+        stop = true;
+        return group_res.status();
+      }
+      if (group = group_res->second; !group) {
+        (*cntx)->SendError(NoGroupError(key, opts.gname));
+        stop = true;
+        return OpStatus::SKIPPED;
```

Comment:
To follow up:

returning a status is relevant only for single-shard transactions like GET/SET etc.

here it's a quite complicated multi-shard transaction so returning a status via return value is not relevant (this lambda runs multiple times). Instead, we use `res_sitems` to track a status per shard. It should be   `vector<OpResult<SomeValue>>` where `SomeValue` in this case is `unordered_map<...>`.

you also should not use `stop` variable here because it will be written concurrently by multiple threads. Instead, when the first home (execute call) finishes, you can observe the status of each result in   `res_sitems` and act accordingly.